### PR TITLE
New Datapath Extension: Broadcaster

### PR DESCRIPTION
**Pull request by IveanEx**
This PR introduces the new DataPath extension: by using this extension, C channel of streamer can reduce the unnecessary fetches from the reader. 

In this PR, we also add the broadcast C extension flag, especially for the gemm C32 readers. With this extension enabled, every output channel can share one bias. More specifically, we add:
* streamer with the has_C_broadcast flag support to enable the broadcast extension, none by default
* csr modification for the broadcast C extension, relevant sw test and snaxgen

### CODE REVIEW

**Review by xiaoling-yi — APPROVED** [review]

File: hw/chisel/src/main/scala/snax/streamer/DesignParams.scala [hunk 1]
```diff
@@ -22,6 +22,9 @@ class StreamerParam(
     // transpose params
     val hasTranspose: Boolean = false,
 
+    // c broadcast params
+    val hasCBroadcast: Boolean = false,
+
```
Comment (jorendumoulin) [inline hw/chisel/src/main/scala/snax/streamer/DesignParams.scala:27]: why not just `hasBroadcast`? I feel like this extension could be used in other places than gemm as well?

File: hw/chisel/src/main/scala/snax/streamer/DesignParams.scala [hunk 2]
```diff
@@ -86,13 +89,28 @@ class StreamerParam(
...
+  val dataPathABExtensionParam: Seq[HasDataPathExtension] =
     (if (hasTranspose)
        Seq[HasDataPathExtension](
          HasTransposer
        )
      else
        Seq[HasDataPathExtension]())
+
+  val dataPathCExtensionParam: Seq[HasDataPathExtension] =
+    (if (hasCBroadcast)
```
Comment (jorendumoulin) [inline hw/chisel/src/main/scala/snax/streamer/DesignParams.scala:101]: these are very gemm-specific parameters in the main streamer files, can we make this a bit more generic?

**Review by jorendumoulin — APPROVED** [review]
Thank you, looks good!

So to program this, I need to disable all but the first channel, and then enable the broadcast extension? Is that right?

Something that's not very elegant is there is a lot of specific gemmx-streamer code in the general streamer code. Could we make this a bit more generic in the future?
I feel like if we specify the enabled extensions for every streamer separately in the `cfg/*.hjson` file instead of for the entire accelerator, this would make much more sense.

**Comment by IveanEx** [comment]
> Thank you, looks good!
> 
> So to program this, I need to disable all but the first channel, and then enable the broadcast extension? Is that right?
> 
> Something that's not very elegant is there is a lot of specific gemmx-streamer code in the general streamer code. Could we make this a bit more generic in the future? I feel like if we specify the enabled extensions for every streamer separately in the `cfg/*.hjson` file instead of for the entire accelerator, this would make much more sense.

You have to enable the first eight channels in the system, as the first eight channels represent the first "wide" channel (although only 256bits are used). This block is connected to the wide port.

File: hw/chisel/src/main/scala/snax/streamer/DesignParams.scala (hunk 1 above)
Comment (xiaoling-yi) [reply hw/chisel/src/main/scala/snax/streamer/DesignParams.scala:27]: In theory yes, but actually, only the gemm's streamer uses extensions, and the extensions are also designed for gemm (specific parameters etc).
We can improve it to look more generic later though.

**Pull request merged**